File: sync/src/main/java/com/inceptionnotes/sync/ws/WebsocketClient.java
package com.inceptionnotes.sync.ws;


import com.inceptionnotes.sync.Client;
import com.inceptionnotes.sync.Server;

import java.io.IOException;
import java.util.logging.Logger;

import javax.websocket.EndpointConfig;
import javax.websocket.OnClose;
import javax.websocket.OnError;
import javax.websocket.OnMessage;
import javax.websocket.OnOpen;
import javax.websocket.Session;
import javax.websocket.server.ServerEndpoint;

import static java.util.concurrent.TimeUnit.MINUTES;

@ServerEndpoint(value = "/ws", configurator = WebsocketServer.class)
public class WebsocketClient {

    private Session session;
    private Server server;
    private Client client;

    @OnOpen
    public void onOpen(Session session, EndpointConfig endpointConfig) {
        Logger.getAnonymousLogger().info("WEBSOCKET (SESSION): " + session.getId());
        this.session = session;

        session.setMaxIdleTimeout(MINUTES.toMillis(30));
        server = (Server) endpointConfig.getUserProperties().get("server");
        client = new Client(this);
        server.join(client);
    }

    @OnClose
    public void onClose() {
        if (session.isOpen()) {
            try {
                session.close();
            } catch (IOException | IllegalStateException e) {
                e.printStackTrace();
            }
        }

        Logger.getAnonymousLogger().info("WEBSOCKET: END");
        server.leave(client);
        client.close();
    }

    @OnMessage
    public void onMessage(String message) throws IOException {
        Logger.getAnonymousLogger().info("WEBSOCKET (MESSAGE): " + (message.length() > 128 ? message.substring(0, 127) + "..." + message.length() : message));
        client.got(message);
    }

    @OnMessage
    public void onData(byte[] data) throws IOException {
        Logger.getAnonymousLogger().info("WEBSOCKET (DATA): " + data.length);
        client.got(data);
    }

    @OnError
    public void onError(Throwable t) throws Throwable {
        Logger.getAnonymousLogger().info("WEBSOCKET (ERROR): " + t.getMessage());
        t.printStackTrace();
    }

    public Session getSession() {
        return session;
    }

    public Client getClient() {
        return client;
    }

    public Server getServer() {
        return server;
    }
}

File: sync/src/main/java/com/inceptionnotes/sync/ws/WebsocketClient.kt
package com.inceptionnotes.sync.ws


import com.google.gson.JsonArray
import com.inceptionnotes.sync.events.ServerEvent
import com.inceptionnotes.sync.util.Events
import com.inceptionnotes.sync.util.HttpTransport
import com.inceptionnotes.sync.util.Json
import com.inceptionnotes.sync.util.Outbox
import com.inceptionnotes.sync.world.Client
import com.inceptionnotes.sync.world.Server
import java.io.IOException
import java.util.concurrent.TimeUnit.MINUTES
import javax.websocket.*
import javax.websocket.server.ServerEndpoint

@ServerEndpoint(value = "/ws", configurator = WebsocketServer::class)
class WebsocketClient {

    lateinit var session: Session
        private set
    lateinit var server: Server
        private set
    lateinit var client: Client
        private set

    private var clientToken: String? = null
    private val outbox = Outbox()

    @OnOpen
    fun onOpen(session: Session, endpointConfig: EndpointConfig) {
        this.session = session

        session.maxIdleTimeout = MINUTES.toMillis(5)
        server = endpointConfig.userProperties["server"] as Server
        client = Client(server.on, {
            clientToken = it
            server.on<HttpTransport>().register(it, this::onHttpMessage)
            client.send(ServerEvent("identified"))
        }) {
            outbox.add(it)
            flushOutbox()
        }

        server.join(client)
        client.open()
    }

    private fun onHttpMessage(message: String): String {
        val events = Json.json.fromJson(message, JsonArray::class.java)

        for (event in events) {
            client.got(Json.json.fromJson(
                    event.asJsonArray.get(1),
                    Events.events[event.asJsonArray.get(0).asString]
            ))
        }

        val response = outbox.json()
        outbox.clear()
        return response
    }

    private fun flushOutbox() {
        try {
            val outboxString = outbox.json()

            if (outboxString.length < 1000) {
                session.basicRemote.sendText(outboxString)
                outbox.clear()
            } else {
                requestHttp()
            }
        } catch (ex: IOException) {
            ex.printStackTrace()
        } catch (ex: IllegalStateException) {
            ex.printStackTrace()
        }
    }

    private fun requestHttp() {
        val events = JsonArray()
        val event = JsonArray()
        val serverEvent = ServerEvent("fetch")
        event.add(Events.actions[serverEvent::class.java])
        event.add(Json.json.toJsonTree(serverEvent))
        events.add(event)
        session.basicRemote.sendText(Json.json.toJson(events))
    }

    @OnClose
    fun onClose() {
        if (session.isOpen) try {
            session.close()
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: IllegalStateException) {
            e.printStackTrace()
        }

        clientToken?.let { server.on<HttpTransport>().unregister(it) }

        server.leave(client)
        client.close()
    }

    @OnMessage
    @Throws(IOException::class)
    fun onMessage(message: String) {
        val events = Json.json.fromJson(message, JsonArray::class.java)

        for (event in events) {
            client.got(Json.json.fromJson(
                    event.asJsonArray.get(1),
                    Events.events[event.asJsonArray.get(0).asString]
            ))
        }
    }

    @OnMessage
    @Throws(IOException::class)
    fun onData(data: ByteArray) {
        // Do nothing
    }

    @OnError
    @Throws(Throwable::class)
    fun onError(t: Throwable) {
        t.printStackTrace()
    }
}